src/main/scala/mesosphere/marathon/core/deployment/impl/TaskReplaceActor.scala

```diff
@@ -74,189 +58,331 @@ class TaskReplaceActor(
     eventBus.subscribe(self, classOf[InstanceChanged])
     eventBus.subscribe(self, classOf[InstanceHealthChanged])
 
-    // reconcile the state from a possible previous run
-    reconcileAlreadyStartedInstances()
+    // Kill instances with Goal.Decommissioned. This is a quick fix until #6745 lands.
+    val doomed: Seq[Instance] = instances.valuesIterator.filter { instance =>
+      (instance.state.goal == Goal.Decommissioned || instance.state.goal == Goal.Stopped) && !considerTerminal(instance.state.condition)
+    }.to[Seq]
+    killService.killInstancesAndForget(doomed, KillReason.Upgrading)
 
     // kill old instances to free some capacity
-    for (_ <- 0 until ignitionStrategy.nrToKillImmediately) killNextOldInstance()
-
-    // start new instances, if possible
-    launchInstances().pipeTo(self)
+    logger.info("Sending immediate kill")
+    self ! KillImmediately(ignitionStrategy.nrToKillImmediately)
 
     // reset the launch queue delay
     logger.info("Resetting the backoff delay before restarting the runSpec")
     launchQueue.resetDelay(runSpec)
-
-    // it might be possible, that we come here, but nothing is left to do
-    checkFinished()
   }
 
   override def postStop(): Unit = {
     eventBus.unsubscribe(self)
     super.postStop()
   }
 
-  override def receive: Receive = initializing
+  override def receive: Receive = killing
+
+  // Commands
+  case object Check
+  case object KillNext
+  case class KillImmediately(oldInstances: Int)
+  case class Killed(id: Seq[Instance.Id])
+  case object ScheduleReadiness
+  case object LaunchNext
+  case class Scheduled(instances: Seq[Instance])
+
+  /* Phases
+  We cycle through the following update phases:
+
+  1. `updating` process instance updates and apply them to our internal state.
+  2. `checking` Check if all old instances are terminal and all new instances are running, ready and healthy.
+  3. `killing` Kill the next old instance, ie set the goal and update our internal state. We are ahead of what has been persisted to ZooKeeper.
+  4. `launching` Scheduler one readiness check for a healthy new instance and launch a new instance if required.
+  */
+
+  def updating: Receive = {
+    case InstanceChanged(id, _, _, _, inst) =>
+      logPrefixedInfo("updating")(s"Received update for ${readableInstanceString(inst)}")
+      // Update all instances.
+      instances = instanceTracker.specInstancesSync(runSpec.id).map { i => i.instanceId -> i }(collection.breakOut)
```

Comment:
as we discussed on friday, I would probably go the way of applying condition updates from `InstanceChanged` instead of pulling/overriding the whole instance